### DISCUSSION TITLE
Suggested path to a 5.0 RC

### DISCUSSION THREAD

**Discussion by joelostblom** [vega/altair#2599] [General]
@jakevdp @ChristopherDavisUCI @mattijn I made this so that it is easier for us to keep track of what needs to be done before a release candidate for Altair 5 is published. This is just my understanding of what is left from skimming the open issues, so feel free to suggest changes. Once we have a release candidate out, I can go through all open issues in more detail to close everything that works in the RC.

## Important

I think we should at minimum address these five issues before a release candidate is created.

- [x] https://github.com/altair-viz/altair/pull/2566 (merged)
- [x] https://github.com/altair-viz/altair/pull/2577 (merged)
- [x] https://github.com/altair-viz/altair/pull/2702 (merged)
- [x] https://github.com/altair-viz/altair/pull/2582 (merged)
- [x] https://github.com/altair-viz/altair/pull/2576 (merged)
- [x] https://github.com/altair-viz/altair/issues/2597 (merged)
- [x] https://github.com/altair-viz/altair/issues/2703 (merged)
- [x] https://github.com/altair-viz/altair/issues/2836 (merged)
- [x] https://github.com/altair-viz/altair/issues/3030 (merged)
- [x] https://github.com/altair-viz/altair/issues/2849 (merged)

## Dream come true

Technically not blocking but I am :heart_eyes: for this issue. I think it is more attractive than the "nice-to-haves", and more important to make a decision on before the 5.0 release since it would be preferred to introduce such as big addition in a major release.

- [x] https://github.com/altair-viz/altair/pull/2795 (merged)

## Nice to have

These are not blocking and could be part of minor 5.x releases.

- [x] https://github.com/altair-viz/altair/pull/2625 (merged)
- [x] https://github.com/altair-viz/altair/issues/2581 (merged)
- [x] https://github.com/altair-viz/altair/issues/2497 (merged)
- [x] https://github.com/altair-viz/altair/pull/2568 (merged)
- [x] https://github.com/altair-viz/altair/pull/2522 (merged)
- [x] https://github.com/altair-viz/altair/pull/2814 (merged)
- [x] https://github.com/altair-viz/altair/pull/2842 (merged)
- [x] https://github.com/altair-viz/altair/issues/2908 (merged)
- ~https://github.com/altair-viz/altair/pull/2835 (not planned)~
- ~https://github.com/altair-viz/altair/issues/2584 (not planned)~
- ~https://github.com/altair-viz/altair/pull/2586 (not planned)~
- [x] https://github.com/altair-viz/altair/issues/2867 (merged)
- [ ] https://github.com/altair-viz/altair/issues/2820 (not started)

Docs (these could come at a later point as well, but would be great to sync with a big release):
- [x] https://github.com/altair-viz/altair/issues/2580 (merged)
- [x] https://github.com/altair-viz/altair/issues/2578 (merged)
- [x] https://github.com/altair-viz/altair/issues/2572 (merged)
- [x] https://github.com/altair-viz/altair/pull/2415 (merged)

## Downstream dependent packaging issues

- [x] https://github.com/altair-viz/altair/issues/2865 (done) [altair-viz-v4]
- [x] https://github.com/altair-viz/altair_viewer/issues/52 (done) [altair_viewer]
- [x] https://github.com/altair-viz/altair/issues/2585 (not started) [jupyter-lab]
- ~https://github.com/altair-viz/altair/issues/2866 (not started) [altair_viewer & altair_saver]~

## Upstream dependent VL issues 

The VL-specification are created like we anticipated for the compound charts in Altair, but there are a few outstanding issues for rendering on the VL-side:

- [x] https://github.com/vega/vega-lite/issues/7854
- [x] https://github.com/vega/vega-lite/issues/8523
- [x] https://github.com/vega/vega-lite/issues/8348
- [x] https://github.com/vega/vega-lite/issues/8446

```mermaid
graph LR;
  vega-lite:::doneok--"Open VL issues (0)" -->altair
  subgraph altair
    svlg--"Open issues (1)"--> slvg_done 
    thef--"Open issues (0)"--> thef_done:::doneok
    ds--"Open issues (0)"--> ds_done:::doneok 
    direction LR
    subgraph svlg ["support VL grammar"]
    end
    subgraph slvg_done ["all done"]
    end
    subgraph thef ["type hinting & error feedback"]
    end
    subgraph thef_done ["all done"]
    end
    subgraph ds ["new docs & site"]
    end
    subgraph ds_done ["all done"]
    end
  end
  subgraph altair-viz-v4 ["altair-viz-v4 \narchived v4 docs"]
    end
  altair--"Open issues (0)"-->altair_viewer:::doneok
  altair--"Open issues (1)"-->altair_saver
  altair--"Open issues (0)"-->jupyter-lab:::doneok
  altair--"Open issues (0)"-->altair-viz-v4:::doneok
  classDef doneok stroke:#3C793E,fill:#53A158,fill-opacity:0.5
```

**Comment by joelostblom** [comment]
@jakevdp  is there anything I can do to help with the review of the open PRs and the progress towards the release of the next Altair version? I would be happy to co-maintain the library if you are comfortable with me reviewing and merging low risk PRs, such as simpler fixes (e.g. #2576, #2577, #2582) and those related to the docs (e.g. #2566, #2607, and #2610).I think we have some momentum currently with multiple people contributing and it would be great if we could sustain that as much as possible while not overloading you with things to review, but I understand that this can be a big decision so I am happy to listen to other ideas of how I can be helpful here. 

@mattijn  maybe you would also be interested to help out with PR review/merging too if that is an option?

**Reply by mattijn** [reply]
Yes, yes!
Would be a pity if the master branch get stale and simultaneously too much adopted in current condition.

**Reply by ChristopherDavisUCI** [reply]
That sounds like a great idea to me, if @joelostblom and @mattijn could help to merge some of the more modest PRs.  (I'd offer to help myself but I don't understand git/GitHub well enough.)  I'm very interested in working on some updates (especially over the summer) but I've found it difficult to keep track of the not-yet-merged work.  In general @jakevdp if there's anything I can do to help, please just let me know and I'll be very happy to try.  (As a side note, I'm excited to see new people contributing recently!)

**Comment by mattijn** [comment]
As was pointed out in https://github.com/altair-viz/altair/pull/2686#issuecomment-1299080517. Once Altair 5 RC is released `altair_viewer` and `altair_saver` won't work anymore as they are using the entrypoints of altair vegalite 4.

Observe current behaviour using altair from GitHub main branch:
```python
import altair as alt
from vega_datasets import data
import altair_viewer

cars = data.cars.url

op_var = alt.param(value=0.1)

chart = alt.Chart(cars).mark_circle(opacity=op_var).encode(
    x='Miles_per_Gallon:Q',
    y='Horsepower:Q',
    color='Origin:N'
).add_params(
    op_var
)

altair_viewer.display(chart)
```

> ```cmd
> ---------------------------------------------------------------------------
> NoMatchingVersions: No matches for version='5.2.0' among ['4.0.2', '4.8.1', '4.17.0'].
> Often this can be fixed by updating altair_viewer:
>     pip install -U altair_viewer
> ```

**Reply by mattijn** [reply]
I just noticed the last checkbox in the OP, that describes already that these companion packages require updating.

**Comment by ChristopherDavisUCI** [comment]
@joelostblom @mattijn Is there anything in particular that I could focus on to help with this release?

**Reply by binste** [reply]
Same question, how can I best help? :) I'm currently working on updating and simplifying the README.md file (will create a PR soon). Furthermore, the following PRs are open from my side and ready for review: #2755, #2758, #2771.

**Reply by mattijn** [reply]
I will have a look to some of the outstanding PRs in a few hours. 

I would favor to archive the current released documentation of https://github.com/altair-viz/altair-viz.github.io in a v4 repository. Maybe just like https://github.com/vega/vega-lite-v4? 

The docs are currently synchronized from here: https://github.com/altair-viz/altair/blob/master/doc/sync_website.sh, but it's not entirely clear to me how we can approach this best. Could you have a look @ChristopherDavisUCI? 

@binste, could you please have a look to this issue: https://github.com/altair-viz/altair/issues/2761, the data section is currently split in different pages, but it is not yet filled in and the main landing page is currently still https://github.com/altair-viz/altair/blob/master/doc/user_guide/data.rst, which require some changes to accommodate.

**Reply by binste** [reply]
Thanks mattjin for the feedback here and in the open PRs and issues! I'll have a look.

I'm also in favor of archiving the v4 docs. For reference, we should add something similar to the "This docs is for version 4" banner at the top of https://vega.github.io/vega-lite-v4/. Code at https://github.com/vega/vega-lite-v4/blob/master/_layouts/plain.html#L7 On the new page we could try to add a [version switcher](https://pydata-sphinx-theme.readthedocs.io/en/stable/user_guide/version-dropdown.html) to make it easy to get to the v4 docs.

**Reply by ChristopherDavisUCI** [reply]
Thanks @mattijn, great to see all your progress!  I took a look at what you linked for the docs, but it's all pretty incomprehensible to me unfortunately.  I assume you've seen this page that gives some (also not very comprehensible to me) instructions related to the docs and releasing: https://github.com/altair-viz/altair/blob/master/RELEASING.md

(I also wasn't sure what specific part you were asking about... was it more about getting v5 published, or more about how to archive v4?)

Maybe @binste would have more helpful feedback on this?

**Reply by binste** [reply]
The altair documentation is hosted using [github pages](https://pages.github.com) as an "Organization site". This means that the repository https://github.com/altair-viz/altair-viz.github.io contains the source files for the website. If you push changes to those files, they are automatically deployed and visible at https://altair-viz.github.io. I'm doing the same to host a development version of the docs at https://binste.github.io/altair-docs/index.html

Before running `sync_website.sh`, one needs to build the documentation locally as described in `RELEASING.md`. Then, the script will:

* Clone https://github.com/altair-viz/altair-viz.github.io
* Replacing the files for the current documentation with the new documentation. You'll need to make sure that the `.nojekyll` file still is in the docs repository after this step as it does not come from sphinx but needs to be added manually. Not sure if the script respects this or will delete it
* Pushing the new files to GitHub where they will be automatically deployed

To create an archived version of the v4 docs, it would probably look something like this:

* Create a new repo called for example "altair-viz-v4"
* Copy over current files from the https://github.com/altair-viz/altair-viz.github.io repo (including the `.nojekyll` file), commit and push
* Go to the repo settings and set up GitHub Pages as described in step 4 for "Project site" [here](https://pages.github.com)
* Page should be live within a few minutes at https://altair-viz.github.io/altair-viz-v4

Hope this is helpful! Happy to assist with any of these steps if needed.

**Reply by mattijn** [reply]
This is really helpful @binste. Thanks for making this really clear. 
@jakevdp, do you agree with this as well? Since we don't have the rights, we were wondering if you could assist us in this process of setting up another repo within the altair-viz organisation page.

**Reply by mattijn** [reply]
Regarding the documentation. I'm fine to publish the v5 altair release page together with the altair 5.0rc1. Maybe we can add an announcement header there as well just like the archive-page of the v4-docs.

**Comment by mattijn** [comment]
@binste, I saw your comment regarding an official Altair release here: https://github.com/altair-viz/altair/pull/2770#issuecomment-1367107039. So I thought I add some more info here: 

Since we have adopted this top-level parameter strategy, there are still a few outstanding issues on the VL side which will currently cause regression on the Altair side. So I think we have to wait for a new release on the VL side (`v5.7.0`?), before we can do an official `v5.0` release of Altair. 

For a release candidate version of Altair (`v5.0rc1`) I'm willing to accept these discrepancies and publish against Vega-Lite`v5.2.0` since that version of Vega-Lite is currently available in the whole Altair family of packages (altair, altair_saver, altair_viewer).

The outstanding VL-issues are currently summarized in this comment https://github.com/altair-viz/altair/pull/2702#issuecomment-1295853109.  
These can be reduced to the following one/two issues, which I think are the main bottleneck:
- https://github.com/vega/vega-lite/issues/7854 (layered chart with point=True, very common syntax)
- https://github.com/vega/vega-lite/issues/8348 (maybe not so common in practice, but one of the main selling-charts of VL)

And since it is currently holiday season we hope that the people involved can find some relaxing code time;)

After there is a new release of Vega-Lite, there are a few repositories within the altair-viz family that require a new release. That will take, with a bit of luck, one or two weeks.

**Reply by binste** [reply]
Thank you for the detailed explanation @mattijn! I wasn't aware of these outstanding issues and now see why it makes sense to wait with the official 5.0 release.

For the new docs, apart from the announcement banner you mentioned, we would probably want to update the installation page with a notice on how to install the pre-release candidate. I think with pip you would either need to do `pip install altair==5.0rc1` or `pip install --pre`. See https://pip.pypa.io/en/stable/cli/pip_install/#pre-release-versions Similar instructions are probably required for conda.

**Reply by joelostblom** [reply]
I am also on board with having a release candidate that still contains these few VegaLite incompatibilities and encourage people to try it out and give feedback. Then we can maybe have time to fix issues from that feedback by the time VegaLite has been updated and we can make a 5.0 release.

Regarding the installation command I am OK to just mention the pip way for the RC in case it is complicated for conda (we could even just link to the github tag/commit if we don't want to / have issue with publishing to PyPI). In addition to mentioning it in the doc page we can also create a pinned issue once the RC is out.

**Reply by binste** [reply]
I'm wondering if any `vega-lite` maintainers are aware of which issues block the Altair 5 release. I don't want to pressure anyone, just thinking about creating an issue in the `vega-lite` repo which summarises the open points so it's easier to track for them and hopefully resolve. Are you aware of such a summary? Happy to create one based on the information from @mattijn provided above. Or is anyone in direct contact with a `vega-lite` maintainer?

**Reply by mattijn** [reply]
I've had contact with @arvind before and he is aware. If we can assist directly with the Vega-lite repositroy to resolve these issues then that would be great too and I'm sure much appreciated. Without speaking for anyone, people grow into other roles and are scrutinised in the publish or perish / tenure adventure, together with private obligations, time becomes scare for open-source maintenance work, but simultaneously are much delighted to see our work. 
Shall we give it a try to fix an issue in the Vega-lite repository? It means that we, people with a python background, must go out of our comfort zone. Technically speaking as far as I can see, this issue, https://github.com/vega/vega-lite/issues/7854, should not be super complicated to start with, but I agree, this is upstream work, and always it had been a nice fence, your side/typescript, our side/python. But I'm sure with a few pointers we can get things moving a bit to unclog these issues as well. And without getting anyone involved, maybe @ChristopherDavisUCI is also open to give this a try? What do you think @binste?

**Reply by ChristopherDavisUCI** [reply]
That sounds very good to me @mattijn, thanks!  I got a Vega-Lite development environment working once before (maybe a year ago) and will start out trying to get back to that stage.

**Reply by binste** [reply]
Sure, totally understand that! And I like the idea of building up the skills to contribute some fixes in the Vega-lite repo ourselves and tear down that fence, I'm in :) I have only some minor javascript experience but will try to brush it up over the coming days and then join the discussion on https://github.com/vega/vega-lite/issues/7854.

**Reply by joelostblom** [reply]
I think a summary of the outstanding VL issues that a new Altair release is severely affected by would be great and I edited my original post above to include those mentioned here https://github.com/altair-viz/altair/pull/2702#issuecomment-1295853109 (let me know if there are others). I believe it could be helpful to dedicate some effort toward fixing these issues in the Vega-Lite repo, but that we should make sure to prioritize things in the Altair repo still. I am personally unsure about contributing directly to VL at this point in time because I think most of my time will be spent being stuck on simple JS things and environment setup... =/

My understanding of the current VL issues is that none of them is so severe that they would block creating a release candidate without this functionality as long as it is clearly stated in the release notes (which we should create btw #2836). Some of them I think could even be included in a point release of Altair, but let me know if you agree/disagree. Other than that, if anyone of you has the time to drop some review comments on any of these PRs: #2568, #2522, #2574, #2835 that would be great because I think those are all something we would want in an RC to get feedback on.

**Comment by ChristopherDavisUCI** [comment]
Hi @mattijn and @joelostblom, is there anything I can help with here?  Thanks for all your work!

**Reply by joelostblom** [reply]
Thank **you**! If you would like to work on https://github.com/altair-viz/altair/issues/2703 I think that could be helpful. It is possible that we already are testing most/all of that, but we should go through the Vega-Lite release notes and make sure that we have a gallery examples (or other tests) that covers all the new functionality.

My thoughts in general for getting a new release out while not having release permissions for the time being, are that the best we can do is to create a tag on GitHub that we name `v5.0.0rc1` and encourage users to install and try it out. We could create a temporary site for the new documentation and make an announcement in an issue that shows how to install the GitHub tag and links to the documentation (and mentions the caveats for altair_saver and altair_viewer).

If nothing major comes up after a few weeks of receiving feedback from users, we can reach out to Jake again saying that we are ready to make a release, but hopefully he will be able to get back to us during the time we are collecting feedback from the RC. cc @binste too in case you don't get a notification otherwise (not sure how they work in "discussions")

**Reply by binste** [reply]
Agree, also can't think of a better way until you get the relevant permissions. We could also create a PR for the existing docs to addan announcement banner linking to the alternative docs for the RC. Maybe Jake sees the PR. (Appreciate the CC, I get notifications for all replies in this discussion anyway, I think it's similar to Issues 🙂)

**Reply by joelostblom** [reply]
@ChristopherDavisUCI Another issue that I think would be quite helpful to fix is https://github.com/altair-viz/altair/issues/2744. Not crucial for a release candidate, but it makes some specs very hard to troubleshoot since the wrong error message is shown.

**Reply by joelostblom** [reply]
Following up on this, @mattijn and I just had a chat with the Vega-Lite developers and they recommended Altair to be updated to Vega-Lite 5.6 before publishing our first release candidate and that this update should be straightforward as there hasn't been any major new functionality introduced since 5.2. They will then try to work on the issues blocking Altair and make a 5.7 release with those so that we can update to 5.7 before the actual release of Altair 5.

@ChristopherDavisUCI I think you are the most familiar with the update process after all your previous work on 4.17 and 5.2, so I want to ask if you think you would have to work on this, but I also know that you are contributing a lot right now including upstream work in  vega-lite and jupyterlab so I understand that you might not have any free cycles at the moment.

Another strategy here would be to wait until 5.7 is release and just upgrade from 5.2 to 5.7 right away and not have to do two separate PRs. This depends on how quickly Vega-Lite will release 5.7 and we don't want to wait if this delays the release candidate significantly.

**Reply by ChristopherDavisUCI** [reply]
Thanks @joelostblom, that sounds great to me and updating to 5.6 (and later 5.7) should be straightforward.  I will work on this soon.

**Reply by ChristopherDavisUCI** [reply]
Hi @joelostblom, I tried updating to Vega-Lite 5.6.0 here: #2871

**Reply by ChristopherDavisUCI** [reply]
Hi @joelostblom, I'm slowly looking through some of the issues in your original list above.  I'd personally vote for moving `Support Vega-Lite's optional encoding types` https://github.com/altair-viz/altair/issues/2584 to its own "Not planned for 5.0" section (or just doing a ~~strikethrough~~ through it).  Even though it would be easy to implement (I think), I don't find the default encoding types chosen by Vega-Lite to be very compatible with the current defaults inferred from pandas DataFrames.

Longterm I like the idea of taking an approach like @mattijn mentioned here https://github.com/altair-viz/altair/issues/2868#issuecomment-1418243196, but I don't see a need to try to rush to pursue that for the next release.

I'm happy to hear any thoughts on this!

**Reply by mattijn** [reply]
Done

**Reply by ChristopherDavisUCI** [reply]
I'm amazed at all the activity the last few days @mattijn @joelostblom @binste!  Is there anything in particular I should look at?  I have some time today (but then not again until Thursday).

**Reply by mattijn** [reply]
I was thinking, we still should add a section of the limitations of rc1, when not yet try and what things we know off that will not work yet.

**Reply by joelostblom** [reply]
Yup I agree with adding the limitations. What do you think of creating an issue when we release the RC and linking to it in the banner of the docs and pinning it to the repo? The issue could read:

> ## Altair 5 release candidate info and feedback
> 
> Thanks for trying out the release candidate for Altair 5! If you have any general feedback on the direction of the release candidate, feel free to share it here. If you run into specific problems, please open individual issues for those.
>
> Before you get started, make sure that you read the [release notes] to see what has changed, and that you are aware of the following limitations of the release candidate that we hope will be resolved by the time Altair 5.0 is ready to be released:
> 
> - Offline rendering in JupyterLab, using `alt.renderers.enable('mimetype')`, is currently awaiting an updated renderer in JupyterLab.  We expect this to be available soon, in particular, when the PR https://github.com/jupyterlab/jupyterlab/pull/13893 appears in a JupyterLab release (likely 4.0).
> - Altair 5 will be released with VL5.7, but the 5rc1 is built against VL5.6.1. Meaning that if you are dependent on altair_viewer you cannot update yet, as we plan to release altair_viewer with VL5.7 and not VL5.6.1 (note that altair_saver has altair_viewer as dependency, and that there is a new independent default saving mechanism in Altair 5).
> - https://github.com/vega/vega-lite/issues/7854 (will be fixed at release with VL 5.7)
> - https://github.com/vega/vega-lite/issues/8523
> - https://github.com/vega/vega-lite/issues/8348
> - https://github.com/vega/vega-lite/issues/8446

**Reply by mattijn** [reply]
Also, we should mention that Altair 5 will be released with VL5.7 and that this Altair 5rc1 is build against VL5.6.1. Meaning that if you are dependent on altair_viewer you cannot update yet, as we plan to release altair_viewer with VL5.7 and not VL5.6.1. I think altair_saver has altair_viewer as dependency?

VL issue 7854 is resolved but will be released with VL5.7 so it is still a very important missing feature (maybe the most critical).

**Reply by mattijn** [reply]
Also, the relevant VL issue/spec is interesting, but I think it’s more relevant to have the Altair spec that one cannot use yet.

**Reply by ChristopherDavisUCI** [reply]
@joelostblom Do you think it's worth mentioning that offline rendering won't work in JupyterLab until https://github.com/jupyterlab/jupyterlab/pull/13893 appears in a JupyterLab release?  I believe this will happen in the next JupyterLab release, but it's possible it will not be until version 4.0.   Maybe it could be phrased something like this?

> Offline rendering in JupyterLab, using `alt.renderers.enable('mimetype')`, is currently awaiting an updated renderer in JupyterLab.  We expect this to be available soon, in particular, when the PR https://github.com/jupyterlab/jupyterlab/pull/13893 appears in a JupyterLab release.

**Reply by joelostblom** [reply]
Great ideas! I updated the message above.

**Comment by joelostblom** [comment]
@mattijn @ChristopherDavisUCI @binste I just noticed this issue in JupyterLab repo https://github.com/jupyterlab/jupyterlab/issues/9929 mentioning that the Vegaj-Lite binaries are fixed at version 4 and it is not looking like anyone is working on updating them for the time being. Would this cause issues for the Altair renderers that render offline? And would this include any common renderers like the `jupyterlab` one maybe?

**Reply by ChristopherDavisUCI** [reply]
Hi @joelostblom, for whatever it's worth, I've never successfully gotten offline rendering to work in JupyterLab!  (I asked about this fairly recently, and Jake replied that it's common for people to struggle with this, but I can't track down that comment at the moment.)  I also don't think I've gotten it to work in Jupyter Notebook, but I haven't tried as hard.

**Reply by binste** [reply]
Not fully sure but isn't this related to #2585? Also the comment by Jake which @ChristopherDavisUCI  mentions is probably this one https://github.com/altair-viz/altair/issues/2797#issuecomment-1373712263. So basically a frontend needs to upgrade to vega-lite 5 so that the mimetype renderer and therefore offline rendering works and there is nothing on the Altair side we could do (except for creating a PR in the JupyterLab repo).

**Reply by ChristopherDavisUCI** [reply]
Thanks @binste, that's exactly the comment I meant.  Thanks also for pointing out #2585, that does seem very relevant.

**Reply by ChristopherDavisUCI** [reply]
Is it clear what would need to be done to update things on the JupyterLab side?  I'm happy to try working on that.  https://github.com/jupyterlab/jupyterlab/issues/9929#issuecomment-1404697226 makes it sound like a PR from us would be welcome.

**Reply by joelostblom** [reply]
@ChristopherDavisUCI I think you saw it already, but just in case, this post links older PRs doing the same update from dmoritz https://github.com/jupyterlab/jupyterlab/issues/9929#issuecomment-1407733784

**Reply by ChristopherDavisUCI** [reply]
Thanks @joelostblom, yes, I saw it!  I will take a look soon.  If it seems too far out of my comfort zone, I'll also let you know.

**Comment by jbloom** [comment]
@joelostblom, I'm super excited to see all the activity on the GitHub repo from you and others. As an avid user of `altair` I'm just very curious: what do you expect as likely timeframe to next release?

**Reply by joelostblom** [reply]
I'm happy to hear you are enjoying all the progress made here lately @jbloom, so am I! We want to have a release candidate "soon", but we're currently waiting to see if some of the upstream issues in Vega-Lite can be fixed before we proceed and publish a release candidate. You can view all the unchecked issues in the original post above to get a rough sense of how close we are to a release / release candidate.

**Reply by mattijn** [reply]
What would you think of doing a feature freeze at February 26 and around March 1 release of v5.0.0rc1?

**Reply by ChristopherDavisUCI** [reply]
Sounds exciting to me @mattijn!

**Reply by joelostblom** [reply]
Those dates sound good to me. I won't be able to work much March 6-14, so releasing before then is desirable. From my end, I would like to see a resolution to https://github.com/altair-viz/altair/pull/2568, https://github.com/altair-viz/altair/pull/2824, https://github.com/altair-viz/altair/pull/2892, and https://github.com/altair-viz/altair/pull/2871 before the first RC

**Reply by binste** [reply]
Awesome! 🎉  I created a few more mostly smaller PRs today which would be great to get merged. Open to review other PRs or work on open issues to get everything in until February 26, just ping me.

**Comment by ChristopherDavisUCI** [comment]
Has anyone checked out the new pandas release candidate https://github.com/pandas-dev/pandas/releases/tag/v2.0.0rc0 to see if it impacts us at all?  The jump from v1 to v2 is exciting and a little scary-sounding.  (It's always possible I misunderstand, and maybe the upcoming release is always labeled this way?)

**Reply by mattijn** [reply]
I've a branch where I tested the following [Github Actions matrix run](https://github.com/mattijn/altair/actions/runs/4255418004) succesfully:
```yml
matrix:
  python-version: ["3.8", "3.9", "3.10", "3.11"]
  jsonschema-version: ["4.0", "4.5", "4.17"]
  pandas-version: ["1.5.3", "2.0.0rc0"]
```
But when on `jsonschema<4` (eg `3.0.0` or `3.2.0`), there is still one test failing, see this [action run](https://github.com/mattijn/altair/actions/runs/4255568522/jobs/7403347537), same as https://github.com/altair-viz/altair/issues/2894, but than as error instead of warning.

And using python 3.7 there I can use only pandas 1.3.7, but in combination with jsonschema 4.17 the [tests are passing](https://github.com/mattijn/altair/actions/runs/4255618554/jobs/7403463798), but there is a deprecation warning (same as https://github.com/altair-viz/altair/issues/2895)

**Reply by binste** [reply]
That error is independent of pandas. The test can only ever pass for jsonschema >= 4. I propose a fix in #2931 so that it should work for all jsonschema versions >= 3.

**Reply by binste** [reply]
Regarding the deprecation warning, I thought I fixed that in #2895 because Python deprecated that interface in Python 3.10. But as Altair relies on the package `importlib_metadata` for Python < 3.8 (so only 3.7 as 3.6 is not supported by Altair), it's not fixed for Python 3.7. Reason is that the `importlib_metadata` package also deprecated the interface and this is not captured by the current logic in Altair. [sqlalchemy also struggled](https://github.com/python/importlib_metadata/issues/298) with this. I'll see how they solved it and will create a PR later today.

**Reply by binste** [reply]
See #2932

**Comment by mattijn** [comment]
I just released a new version of Altair (5.0.0rc1) and updated the website at https://altair-viz.github.io/. You can find the package on PyPI at https://pypi.org/project/altair/5.0.0rc1. I'm not sure if we need to release a candidate on Conda yet. Thank you all!

**Reply by joelostblom** [reply]
Wow this is so fantastic to see! :joy:  Thank you for putting the release together @mattijn and a HUGE thanks to all of you: @mattijn, @ChristopherDavisUCI, and @binste for all the time and effort you have spent to make this possible, both recently and over the past two years :star_struck: :trophy:

**Reply by mattijn** [reply]
I was expecting more incoming issues after releasing 5.0.0rc1, but it seems to go rather smooth (touch wood!). 
Do we still see things that requires a rc2?

**Reply by joelostblom** [reply]
Yes, there haven't been much feedback, I guess most things are still working the same as in v4 and we haven't broken anything major which is positive =) It is hard to say how many people have tried out the RC, [according to these stats](https://pepy.tech/project/altair?versions=5.0.0rc1&display=weekly) there are around 10k downloads, but we don't know how many of those are part of GitHub workflows etc (although I would expect that share to be lower for an RC).

I am still leaning towards that it makes sense to have an RC2, mainly because it would be nice to release it against VL 5.7 when that is out and at least try it out for a while ourselves and see that things work as expected. I wonder if we should also push the new methods syntax more aggressively in the docs to make it more discoverable and potentially get more feedback from others? In my mind it makes sense to have it as the default in the docs at some point since it its more convenient. Other than that I think we have all the *necessary* PRs merged already and some that would be nice to have could also be in a minor release.

**Reply by binste** [reply]
Honestly, I was expecting more issues as well but also my own experience has been very smooth with it so far.

I'd also see VL 5.7 as the only change for which it might be useful to have another RC. Even if it's just out for 1-2 weeks so we can test it ourselves as @joelostblom mentioned. We could also use an RC2 to test and update `altair_viewer` and maybe `altair_saver`. Other then that I'm also not aware of any PRs which still need to be merged before a v5 release.

**Reply by mattijn** [reply]
OK, so lets have another RC2 when we have aligned with VL5.7 when released. 
After using the methods-based syntax, I'm also fine with adopting it as default. It is indeed more convenient and a major release can be a good moment to do this.
Also agree that this is also a good moment to see if we can get altair_viewer and altair_saver up and running again with altair v5. I haven't spend much time in these codebases, so I'm a bit hesitant to start, but we can try at least.

**Reply by binste** [reply]
All changes which are relevant for rc2 are in master as far as I know. No stress at all! Just wanted to ask if there is anything outstanding where I can be of help? I think 1-2 weeks would still be a good length for testing rc2 and then I'm very excited for the official release 🥳  What do you think?

**Reply by mattijn** [reply]
Let's try to do a rc2-release this weekend and if nothing pops up, aim for an official release in the week of 8 may (before jupycon).

**Comment by mattijn** [comment]
I just released the second release candidate version of Altair (5.0.0rc2) and updated the website at https://altair-viz.github.io/. You can find the package on PyPI at https://pypi.org/project/altair/5.0.0rc2. to install: `pip install altair==5.0.0rc2`

**Reply by jonmmease** [reply]
What version of Vega-Lite does 5.0.0rc2 use? I see that Vega-Lite [5.8.0 was released today](https://github.com/vega/vega-lite/releases/tag/v5.8.0), and I'm wondering if we need a new release of vl-convert (with 5.8 support) ASAP.

**Reply by mattijn** [reply]
Ah, nice! I thought that the 5.8 release of VL was dependent on another release of Vega (https://github.com/vega/vega-lite/pull/8857#issuecomment-1524102769, that linked PR in Vega is merged, but not released yet, cc @kanitw).
I think we should aim for VL5.8 once we are clear which Vega version is connected to it.

**Reply by binste** [reply]
It uses 5.7.1. In the spirit of a release candidate, I'd suggest that we stick to that version for the official release in case people have tested against it. And then we could do an Altair 5.1 release soon incl. support for the new VL 5.8 PPI feature. Thoughts?

**Reply by jonmmease** [reply]
I'd lean slightly against releasing the final version with Vega-Lite 5.7 because I'm not clear on how the issue in https://github.com/vega/vega-lite/issues/8841 (which is fixed in 5.8.0) might impact Altair users.

**Reply by mattijn** [reply]
I tend to agree as well, but I like to have a confirmation first, since the test suite depends on `altair_viewer` which includes the VL and Vega versions of the actual releases of Altair.

**Reply by mattijn** [reply]
But if we release against vl5.8 we also should cut a rc3 release soon if we want to release next week an official release.

**Reply by mattijn** [reply]
I've prepared https://github.com/altair-viz/altair_viewer/pull/57 and https://github.com/altair-viz/altair/pull/3037. Once we have confirmation if vega-lite 5.8.0 depends on vega 5.25.0 (@kanitw, @domoritz). I can start working towards releasing rc3.

**Comment by mattijn** [comment]
I just released the third release candidate version of Altair (5.0.0rc3) and updated the website at https://altair-viz.github.io/. You can find the package on PyPI at https://pypi.org/project/altair/5.0.0rc3. to install: `pip install altair==5.0.0rc3`. 

I had some connectivity issues during uploading to PyPI, but everything seems to be uploaded fine.

**Reply by mattijn** [reply]
I will start making the v5.0.0 release. Hopefully done within one hour

**Reply by binste** [reply]
This is very exciting 🥳 Thank you @mattijn!
